### PR TITLE
{Core} Improve autocomplete performance by preventing loading all modules

### LINKED ISSUE #24762: Autocomplete is slow when command is not complete
> ### `az feedback` auto-generates most of the information requested below, as of CLI version 2.0.62

**Describe the bug**

If I press TAB after `az account cle`, it takes 4seconds to show suggestion `az account clear`. However `az account -h` only takes 2seconds.

The debug log shows that it reloads modules and extensions when run `az account cle --debug`.
```
cli.azure.cli.core: Could not find a match in the command or command group table for 'account cle'. The index may be outdated.
cli.azure.cli.core: Loading command modules:
```

I think we can make reload rules stricter. When ` _ARGCOMPLETE=1`, if `account cle` is not found, check if `account` is matched. If still not matched, reload.

Related issue: https://github.com/Azure/azure-cli/issues/9157

### PR DESCRIPTION
**Pull request by bebound**
**Description**

Close #24762, related issue #9157, related pr #13294
If the command is not complete, for example: `az account sho`, press TAB, cli loads all modules, which takes 1s. It gets worse if user installs many extensions. This make autocomplete slow.
This PR prevents module loading in autocomplete mode.

If command is `az acc`:
1. module begin with `acc` will be returned in `CommandIndex.get()` .
2. `raw_cmd` is 'acc'. Trim last word, it becomes `''`, return `self.command_table`.
```
cli.azure.cli.core: In autocomplete mode, load command begin with: 'acc'
cli.azure.cli.core: Modules found from index for 'acc': ['azure.cli.command_modules.profile', 'azure.cli.command_modules.resource']

cli.azure.cli.core: Could not find a match in the command or command group table for 'acc'
cli.azure.cli.core: In autocomplete mode, try to match trimmed raw cmd: ''
cli.azure.cli.core: Trimmed raw cmd is empty, return command table.
```

If command is `az account sh`:
1. `account` module is returned in `CommandIndex.get()` .
2. `raw_cmd` is `account sh`. Trim last word, it becomes `account` and is in `self.command_group_table`, return `self.command_table`.
```
cli.azure.cli.core: Modules found from index for 'account': ['azure.cli.command_modules.profile', 'azure.cli.command_modules.resource']

cli.azure.cli.core: Could not find a match in the command or command group table for 'account sh'
cli.azure.cli.core: In autocomplete mode, try to match trimmed raw cmd: 'account'
cli.azure.cli.core: Found a match in the command group table for trimmed raw cmd: 'account'.
```

**Testing Guide**
`az acco` and `az account sh` should be faster.
```
$env:_ARGCOMPLETE=1;
Measure-Command {az account sh}
TotalMilliseconds : 567.0021

Remove-Item Env:\_ARGCOMPLETE;
Measure-Command {az account sh}
TotalMilliseconds : 2935.8341
```

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Core

File: src/azure-cli-core/azure/cli/core/__init__.py [hunk 1]
```diff
@@ -555,6 +568,15 @@ def get(self, args):
...
         index_modules_extensions = index.get(top_command)
+        if not index_modules_extensions and is_autocomplete():
+            # If user type `az acco`, command begin with `acco` will be matched.
+            # If the user just installed a new extension which contains a command named `acco`, he can't get the
+            # expected suggestion. But it's a rare scenario, and it's okay to ignore.
+            logger.debug("In autocomplete mode, load command begin with: '%s'", top_command)
+            index_modules_extensions = []
+            for command in index:
+                if command.startswith(top_command):
+                    index_modules_extensions += index[command]
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/__init__.py:579]: Maybe put these lines after L597 for simplicity?

File: src/azure-cli-core/azure/cli/core/__init__.py [hunk 2]
```diff
@@ -555,6 +570,15 @@ def get(self, args):
         # Check the command index for (command: [module]) mapping, like
         # "network": ["azure.cli.command_modules.natgateway", "azure.cli.command_modules.network", "azext_firewall"]
         index_modules_extensions = index.get(top_command)
+        if not index_modules_extensions and is_autocomplete():
+            # If user type `az acco`, command begin with `acco` will be matched.
+            # If the user just installed a new extension which contains a command named `acco`, he can't get the
+            # expected suggestion. But it's a rare scenario, and it's okay to ignore.
```
Comment (bebound) [inline src/azure-cli-core/azure/cli/core/__init__.py:576]: This does not happen as `CommandIndex().invalidate()` is called after extension installation/remove/update.

File: src/azure-cli-core/azure/cli/core/__init__.py (hunk 1 above)
Comment (bebound) [reply src/azure-cli-core/azure/cli/core/__init__.py:581]: I put it here as the following code convert `index_modules_extensions` to `index_builtin_modules` and `index_extensions`

File: src/azure-cli-core/azure/cli/core/__init__.py [hunk 3]
```diff
@@ -555,6 +569,13 @@ def get(self, args):
         # Check the command index for (command: [module]) mapping, like
         # "network": ["azure.cli.command_modules.natgateway", "azure.cli.command_modules.network", "azext_firewall"]
         index_modules_extensions = index.get(top_command)
+        if not index_modules_extensions and self.cli_ctx.data['completer_active']:
+            # If user type `az acco`, command begin with `acco` will be matched.
+            logger.debug("In autocomplete mode, load command begin with: '%s'", top_command)
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/__init__.py:574]:
```suggestion
            logger.debug("In autocomplete mode, load commands starting with: '%s'", top_command)
```

Let's be consistent with Python's naming `startswith`: https://docs.python.org/3/library/stdtypes.html#str.startswith

File: src/azure-cli-core/azure/cli/core/__init__.py [hunk 4]
```diff
@@ -449,6 +449,19 @@ def _get_extension_suppressions(mod_loaders):
                 if raw_cmd in self.command_group_table:
                     logger.debug("Found a match in the command group table for '%s'.", raw_cmd)
                     return self.command_table
+                if self.cli_ctx.data['completer_active']:
+                    # If the command is not complete in autocomplete mode, we should match shorter command.
+                    # For example, `account sho` should match `account`.
+                    logger.debug("Could not find a match in the command or command group table for '%s'", raw_cmd)
+                    raw_cmd = ' '.join(raw_cmd.split()[:-1])
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/__init__.py:456]: Please avoid reusing existing name. Instead, introduce a new name to make the code easier to read.

**Review by jiasli — APPROVED** [review]

**Pull request merged**